Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -83,6 +84,31 @@ MatrixWorkspace::~MatrixWorkspace() {
   }
 }
 
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::SpectrumNumbers &&spectrumNumbers,
+    Indexing::DetectorIDs &&detectorIDs) {
+  setIndexTranslator(Indexing::IndexTranslator(std::move(spectrumNumbers),
+                                               std::move(detectorIDs)));
+}
+
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::IndexTranslator &&translator) {
+  if (translator.size() != getNumberHistograms())
+    throw std::runtime_error("MatrixWorkspace::setIndexTranslator: Translator "
+                             "size does not match number of histograms in "
+                             "workspace");
+
+  auto &spectrumNumbers = translator.spectrumNumbers();
+  auto &detectorIDs = translator.detectorIDs();
+  for (size_t i = 0; i < getNumberHistograms(); ++i) {
+    auto &spectrum = getSpectrum(i);
+    spectrum.setSpectrumNo(spectrumNumbers[i]);
+    if (!detectorIDs.empty())
```

Comment:
@OwenArnold This is a point that I am not quite happy with: In many places algorithms only update the spectrum numbers, keeping the detector IDs as they were. Here I solve this be setting n `IndexTranslator` with an empty vector of detector IDs. However, it is a bit weird to let `setIndexTranslator` overwrite only parts of the current indices.

What do you think? Is there a way to make this more explicit?
- Should there be two routines, `setSpectrumNumberTranslator` and `setDetectorIDTranslator`?
- Should we just explicitly specify it in the `IndexTranslator` constructor, with some sort of flag (`DetectorIDs::KeepCurrent`)?